README.md
[![Latest Stable Version](http://poser.pugx.org/nguyenanhung/hashids-helper/v)](https://packagist.org/packages/nguyenanhung/hashids-helper) [![Total Downloads](http://poser.pugx.org/nguyenanhung/hashids-helper/downloads)](https://packagist.org/packages/nguyenanhung/hashids-helper) [![Latest Unstable Version](http://poser.pugx.org/nguyenanhung/hashids-helper/v/unstable)](https://packagist.org/packages/nguyenanhung/hashids-helper) [![License](http://poser.pugx.org/nguyenanhung/hashids-helper/license)](https://packagist.org/packages/nguyenanhung/hashids-helper) [![PHP Version Require](http://poser.pugx.org/nguyenanhung/hashids-helper/require/php)](https://packagist.org/packages/nguyenanhung/hashids-helper)

# HashIds Helper

Extends of hashids - Basic, Simple and Lightweight

## Version

- [x] `v1.x` support all PHP version `>=5.6`
- [x] `v2.x` support all PHP version `>=7.0`

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
| ----------- | -------------------- | ---------------- | ------------- |
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |

From Vietnam with Love <3


CHANGELOG.md
# CHANGELOG

## [1.0.0] - 2021/09/24

### Added

- [x] Init Project

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
| ----------- | -------------------- | ---------------- | ------------- |
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |


test/test.php
<?php
require_once __DIR__ . '/../vendor/autoload.php';

use nguyenanhung\Libraries\Hashids\HashIds;

$config = array(
    'salt'          => 'w40):pc6cwS{mn9I_O=B$2Cr;=YXA#',
    'minHashLength' => 8,
    'alphabet'      => 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890'
);

$hash = new HashIds();
$hash->setConfig($config);

$id = 12345679;

echo "Version: " . $hash->getVersion() . PHP_EOL;

echo "RandomId: " . $hash->randomId() . PHP_EOL . PHP_EOL;

echo "SourceId: " . $id . PHP_EOL;
$encodeId = $hash->encodeId($id);
echo "EncodeId: " . $encodeId . PHP_EOL;

$decodeId = $hash->decodeId($encodeId);
echo "DecodeId: " . $decodeId . PHP_EOL;


$id2 = 20210804;
echo "SourceId 2: " . $id2 . PHP_EOL;
$encodeId2 = hashIdsEncode($id2);
echo "EncodeId 2: " . $encodeId2 . PHP_EOL;

$decodeId2 = hashIdsDecode($encodeId2);
echo "DecodeId 2: " . $decodeId2 . PHP_EOL;



src/Project.php
<?php
/**
 * Project hashids-helper
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 09/24/2021
 * Time: 15:18
 */

namespace nguyenanhung\Libraries\Hashids;

/**
 * Interface Project
 *
 * @package   nguyenanhung\Libraries\Hashids
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
interface Project
{
    const VERSION       = '1.0.1';
    const LAST_MODIFIED = '2021-09-25';
    const AUTHOR_NAME   = 'Hung Nguyen';
    const AUTHOR_EMAIL  = 'dev@nguyenanhung.com';
    const AUTHOR_URL    = 'https://nguyenanhung.com';
    const PROJECT_NAME  = 'HashIds Helper';
    const USE_BENCHMARK = false;
    const USE_DEBUG     = false;

    /**
     * Hàm lấy thông tin phiên bản Package
     *
     * @author  : 713uk13m <dev@nguyenanhung.com>
     * @time    : 10/13/18 15:12
     *
     * @return string Current Project Version, VD: 0.1.0
     */
    public function getVersion();
}

helpers/helpers.php
<?php
if (!function_exists('hashIdsEncode')) {
    /**
     * Function hashIdsEncode
     *
     * @param $id
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/24/2021 45:14
     */
    function hashIdsEncode($id): string
    {
        return (new nguyenanhung\Libraries\Hashids\HashIds())->encodeId($id);
    }

}
if (!function_exists('hashIdsDecode')) {
    /**
     * Function hashIdsDecode
     *
     * @param $string
     *
     * @return array|mixed
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/24/2021 53:47
     */
    function hashIdsDecode($string)
    {
        return (new nguyenanhung\Libraries\Hashids\HashIds())->decodeId($string);
    }
}


src/HashIds.php
<?php
/**
 * Project hashids-helper
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 09/24/2021
 * Time: 15:18
 */

namespace nguyenanhung\Libraries\Hashids;

use Exception;
use Hashids\Hashids as BaseHashIds;

/**
 * Class HashIds
 *
 * @package   nguyenanhung\Libraries\Hashids
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
class HashIds implements Project
{
    use Version;

    const SALT       = 'Bear40):pc6cwS{mn9I_O=B$2Cr;=YXA#';
    const MIN_LENGTH = 10;
    const ALPHABET   = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890'; // alphabet

    /**
     * @var array HashIds Config
     * @example https://github.com/nguyenanhung/hashids-helper/blob/main/test/test.php
     * @see     https://github.com/nguyenanhung/hashids-helper/blob/main/test/test.php
     */
    protected $config;

    /**
     * Function setConfig
     *
     * @param $config
     *
     * @return $this
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/24/2021 22:12
     */
    public function setConfig($config): HashIds
    {
        $this->config = $config;

        return $this;
    }

    /**
     * Function getConfig
     *
     * @return array
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/24/2021 27:37
     */
    public function getConfig(): array
    {
        return $this->config;
    }

    /**
     * Function prepareConfig
     *
     * @param $config
     *
     * @return array
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/24/2021 38:44
     */
    protected function prepareConfig($config): array
    {
        if (!is_array($config)) {
            $config = array();
        }
        $salt          = $config['salt'] ?? self::SALT;
        $minHashLength = $config['minHashLength'] ?? self::MIN_LENGTH;
        $alphabet      = $config['alphabet'] ?? self::ALPHABET;

        return array(
            'salt'          => $salt,
            'minHashLength' => $minHashLength,
            'alphabet'      => $alphabet
        );
    }

    /**
     * Function encodeId
     *
     * @param $id
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/15/2021 28:40
     */
    public function encodeId($id): string
    {
        try {
            $config = $this->prepareConfig($this->config);
            $hash   = new BaseHashIds($config['salt'], $config['minHashLength'], $config['alphabet']);

            return $hash->encode($id);
        } catch (Exception $e) {
            if (function_exists('log_message')) {
                log_message('error', $e->getMessage());
                log_message('error', $e->getTraceAsString());
            }

            return $id;
        }
    }

    /**
     * Function decodeId
     *
     * @param $string
     *
     * @return array|mixed
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/15/2021 28:47
     */
    public function decodeId($string)
    {
        try {
            $config = $this->prepareConfig($this->config);
            $hash   = new BaseHashIds($config['salt'], $config['minHashLength'], $config['alphabet']);
            $decode = $hash->decode($string);
            if (empty($decode)) {
                return $string;
            }
            if (count($decode) > 1) {
                return $decode;
            }

            return $decode[0];
        } catch (Exception $e) {
            if (function_exists('log_message')) {
                log_message('error', $e->getMessage());
                log_message('error', $e->getTraceAsString());
            }

            return $string;
        }
    }

    /**
     * Function randomId
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/15/2021 43:30
     */
    public function randomId(): string
    {
        return uniqid('HungNG_', true);
    }
}
